cmd/harmony/main.go

```diff
@@ -296,16 +296,26 @@ func setupNodeAndRun(hc harmonyConfig) {
 	// Prepare for graceful shutdown from os signals
 	osSignal := make(chan os.Signal)
 	signal.Notify(osSignal, os.Interrupt, syscall.SIGTERM)
-	go func() {
+	go func(node *node.Node) {
 		for sig := range osSignal {
 			if sig == syscall.SIGTERM || sig == os.Interrupt {
+				utils.Logger().Warn().Str("signal", sig.String()).Msg("Gracefully shutting down...")
 				const msg = "Got %s signal. Gracefully shutting down...\n"
-				utils.Logger().Printf(msg, sig)
-				fmt.Printf(msg, sig)
+				fmt.Fprintf(os.Stderr, msg, sig)
+				if node.Consensus.Mode() == consensus.Normal {
+					phase := node.Consensus.GetConsensusPhase()
+					utils.Logger().Warn().Str("phase", phase).Msg("[shutdown] commit phase has to wait")
+					maxWait := time.Now().Add(time.Second * 2) // wait up to 2 seconds in commit phase
+					for time.Now().Before(maxWait) &&
+						node.Consensus.GetConsensusPhase() == "Commit" {
```

Comment:
this asynchronous approach can be unreliable for bad timing. Better add synchronous signal to stop new block proposal too (otherwise, new consensus could also be started when the node is killed after the delay).
![image](https://user-images.githubusercontent.com/1042540/101815491-78da9200-3ad4-11eb-8c78-bfdd27566dae.png)

Reply:
This way, can first send signal to stop the new block proposal and then wait for more than 2 seconds (like 5s) for the current consensus to fully finish.

Reply:
good way